File: dart/monosecret/test/conformance_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:monosecret/monosecret.dart';
import 'package:test/test.dart';

void main() {
  final fixtures = _fixturesDirectory();
  final cases = fixtures.listSync().whereType<Directory>().toList()
    ..sort((left, right) => left.path.compareTo(right.path));

  for (final directory in cases) {
    final name = directory.uri.pathSegments
        .where((segment) => segment.isNotEmpty)
        .last;

    test('$name resolve conformance', () async {
      final expected = await _jsonFile(directory, 'expected.json');
      final resolved = await _builder(directory).load();

      try {
        expect(await _canonicalResolved(resolved), expected);
      } finally {
        await resolved.close();
      }
    });

    test('$name no-values conformance', () async {
      final expected = await _jsonFile(directory, 'expected_no_values.json');
      final resolved = await _builder(directory).withNoValues().load();

      try {
        expect(resolved.fields, expected);
      } finally {
        await resolved.close();
      }
    });

    test('$name report conformance', () async {
      final expected = await _jsonFile(directory, 'expected_report.json');
      final report = await _builder(directory).report();

      expect(_canonicalReport(report), expected);
    });
  }

  test('reports typed constraint violations', () async {
    final directory = Directory(
      '${fixtures.parent.path}/constraint-violations',
    );
    final report = await _builder(directory).report();
    final violations = {
      for (final violation in report.constraintViolations)
        violation.kind: violation,
    };

    expect(
      violations.keys,
      unorderedEquals([
        ConstraintViolationKind.atLeastOne,
        ConstraintViolationKind.exactlyOne,
      ]),
    );
    expect(violations[ConstraintViolationKind.atLeastOne]!.group, 'cloud');
    expect(violations[ConstraintViolationKind.atLeastOne]!.present, isEmpty);
    expect(violations[ConstraintViolationKind.exactlyOne]!.group, 'token');
    expect(violations[ConstraintViolationKind.exactlyOne]!.present, [
      'FALLBACK',
      'PRIMARY',
    ]);
  });
}

Directory _fixturesDirectory() {
  var directory = Directory.current.absolute;

  while (true) {
    final fixtures = Directory('${directory.path}/conformance/fixtures');
    if (fixtures.existsSync()) {
      return fixtures;
    }

    final parent = directory.parent;
    if (parent.path == directory.path) {
      throw StateError(
        'Could not find conformance/fixtures from ${Directory.current}.',
      );
    }
    directory = parent;
  }
}

MonosecretBuilder _builder(Directory directory) {
  return Monosecret.builder()
      .withPath('${directory.path}/monosecret.toml')
      .withProvider('dotenv://${directory.path}/.env')
      .withReason('Dart conformance');
}

Future<Map<String, Object?>> _jsonFile(Directory directory, String name) async {
  return (jsonDecode(await File('${directory.path}/$name').readAsString())
      as Map<String, Object?>);
}

Future<Map<String, Object?>> _canonicalResolved(Resolved resolved) async {
  final secrets = <String, Object?>{};

  for (final entry in resolved.secrets.entries) {
    final secret = entry.value;
    final value = secret.asPath
        ? await File(secret.path!).readAsString()
        : secret.value;
    secrets[entry.key] = {
      'value': value,
      'source': secret.source,
      'as_path': secret.asPath,
    };
  }

  return {
    'profile': resolved.profile,
    'secrets': secrets,
    'missing_required': <String>[],
    'missing_optional': [...resolved.missingOptional]..sort(),
  };
}

Map<String, Object?> _canonicalReport(ResolutionReport report) {
  return {
    'profile': report.profile,
    'secrets': {
      for (final secret in report.secrets)
        secret.name: {
          'status': secret.status,
          'required': secret.required,
          'as_path': secret.asPath,
          'generated': secret.generated,
          'default_applied': secret.defaultApplied,
          'source_provider': secret.sourceProvider != null,
        },
    },
  };
}

File: php/monosecret_php/tests/ConformanceTest.php
<?php

declare(strict_types=1);

namespace Monosecret\Tests;

use PHPUnit\Framework\Attributes\DataProvider;
use PHPUnit\Framework\TestCase;
use Monosecret\Builder;
use Monosecret\Report;
use Monosecret\Resolved;
use Monosecret\Monosecret;

/**
 * Cross-language conformance: resolve the shared fixtures and assert this SDK
 * produces the canonical result every other SDK must also produce.
 */
final class ConformanceTest extends TestCase
{
    private const FIXTURES = __DIR__ . '/../../../conformance/fixtures';
    private const CONSTRAINTS = __DIR__ . '/../../../conformance/constraint-violations';

    /** @return iterable<string, array{0: string}> */
    public static function fixtures(): iterable
    {
        foreach (\glob(self::FIXTURES . '/*', \GLOB_ONLYDIR) ?: [] as $dir) {
            yield \basename($dir) => [$dir];
        }
    }

    #[DataProvider('fixtures')]
    public function testConformance(string $dir): void
    {
        $expected = self::readJson($dir . '/expected.json');
        // Close the Resolved so as_path temp files do not accumulate.
        $resolved = $this->builder($dir)->load();
        try {
            self::assertEquals($expected, $this->canonical($resolved));
        } finally {
            $resolved->close();
        }
    }

    /**
     * Under no_values every SDK must emit the same all-null fields map: a
     * value-less secret serializes to null, not an empty string.
     */
    #[DataProvider('fixtures')]
    public function testConformanceNoValues(string $dir): void
    {
        $expected = self::readJson($dir . '/expected_no_values.json');
        $resolved = $this->builder($dir)->withNoValues()->load();
        self::assertEquals($expected, $resolved->fields());
    }

    public function testTypedConstraintViolations(): void
    {
        $report = $this->builder(self::CONSTRAINTS)->report();
        $byKind = [];
        foreach ($report->constraintViolations as $violation) {
            $byKind[$violation->kind->value] = $violation;
        }

        self::assertSame('cloud', $byKind['at_least_one']->group);
        self::assertSame([], $byKind['at_least_one']->present);
        self::assertSame('token', $byKind['exactly_one']->group);
        self::assertSame(['FALLBACK', 'PRIMARY'], $byKind['exactly_one']->present);
    }

    /** The value-free report (status + provenance) is identical across SDKs. */
    #[DataProvider('fixtures')]
    public function testConformanceReport(string $dir): void
    {
        $expected = self::readJson($dir . '/expected_report.json');
        self::assertEquals($expected, $this->canonicalReport($this->builder($dir)->report()));
    }

    private function builder(string $dir): Builder
    {
        return Monosecret::builder()
            ->withPath($dir . '/monosecret.toml')
            ->withProvider('dotenv://' . $dir . '/.env')
            ->withReason('conformance');
    }

    /** @return array<string, mixed> */
    private function canonical(Resolved $resolved): array
    {
        $secrets = [];
        foreach ($resolved->secrets as $name => $secret) {
            $value = $secret->asPath ? \file_get_contents($secret->get()) : $secret->value;
            $secrets[$name] = ['value' => $value, 'source' => $secret->source, 'as_path' => $secret->asPath];
        }
        $missingOptional = $resolved->missingOptional;
        \sort($missingOptional);

        return [
            'profile' => $resolved->profile,
            'secrets' => $secrets,
            'missing_required' => [],
            'missing_optional' => $missingOptional,
        ];
    }

    /** @return array<string, mixed> */
    private function canonicalReport(Report $report): array
    {
        $secrets = [];
        foreach ($report->secrets as $s) {
            $secrets[$s->name] = [
                'status' => $s->status,
                'required' => $s->required,
                'as_path' => $s->asPath,
                'generated' => $s->generated,
                'default_applied' => $s->defaultApplied,
                // Present-or-not (not the path-dependent value) so the vector is
                // machine-independent yet still catches a dropped source_provider.
                'source_provider' => $s->sourceProvider !== null,
            ];
        }

        return ['profile' => $report->profile, 'secrets' => $secrets];
    }

    /** @return array<string, mixed> */
    private static function readJson(string $path): array
    {
        return \json_decode((string) \file_get_contents($path), true, 512, \JSON_THROW_ON_ERROR);
    }
}
